src/process/ProcessManagerImpl.cpp

```diff
@@ -40,20 +41,89 @@
 extern char** environ;
 #endif
 
+namespace
+{
+enum ProcessLifecycle
+{
+    PENDING = 0,
+    RUNNING = 1,
+    TRIED_POLITE_SHUTDOWN = 2,
+    TRIED_FORCED_SHUTDOWN = 3,
+    TERMINATED = 5
+};
+}
+
 namespace stellar
 {
 
 static const asio::error_code ABORT_ERROR_CODE(asio::error::operation_aborted,
                                                asio::system_category());
 
+static asio::error_code
+mapExitStatusToErrorCode(std::string const& cmdLine, int pid, int status)
+{
+// On windows, an exit status is just an exit status. On unix it's
+// got some flags incorporated into it.
+#ifdef _WIN32
+    return asio::error_code(status, asio::system_category());
+#else
+    if (WIFEXITED(status))
+    {
+        if (WEXITSTATUS(status) == 0)
+        {
+            CLOG_DEBUG(Process, "process {} exited {}: {}", pid,
+                       WEXITSTATUS(status), cmdLine);
+        }
+        else
+        {
+            CLOG_WARNING(Process, "process {} exited {}: {}", pid,
+                         WEXITSTATUS(status), cmdLine);
+        }
+#ifdef __linux__
+        // Linux posix_spawnp does not fault on file-not-found in the
+        // parent process at the point of invocation, as BSD does; so
+        // rather than a fatal error / throw we get an ambiguous and
+        // easily-overlooked shell-like 'exit 127' on waitpid.
+        if (WEXITSTATUS(status) == 127)
+        {
+            CLOG_WARNING(Process, "");
+            CLOG_WARNING(Process, "************");
+            CLOG_WARNING(Process, "");
+            CLOG_WARNING(Process, "  likely 'missing command':");
+            CLOG_WARNING(Process, "");
+            CLOG_WARNING(Process, "    {}", cmdLine);
+            CLOG_WARNING(Process, "");
+            CLOG_WARNING(Process, "************");
+            CLOG_WARNING(Process, "");
+        }
+#endif
+        // FIXME: this doesn't _quite_ do the right thing; it conveys
```

Comment:
No, this is the standard exit path (`WIFEXITED`) which gets both error and non-error exits. Non-error is zero which we need to pass through to `asio::error_code` to represent a "successful" exit in callers. Likewise with nonzero (and for the sake of thoroughness we pass the whole nonzero exit value through so users can fish it out with `ec.value`).

The only confusing / awkward part of this is that `asio::error_code` is assumed to transport a `std::error_category`-coded number rather than a process exit code, and so helper methods like `message` do the wrong thing. But there's no other place to _put_ the process exit code. Asio wasn't built assuming it would be conveying process-exits at all. We have to put it somewhere.

Reply:
ah ok - I thought you were saying that we have code paths that call `perror` (or equivalent), so would produce bogus log messages. If that's not the case, it seems fine then.